### ISSUE TITLE
CAS Statistics: Read hit count/ratio is not correct

### ISSUE THREAD

**Issue by eriksun421** [Open-CAS/open-cas-linux#828]
## Description

Read Hit Count and hit ratio is not correct and misleading. 

## Expected Behavior

CAS statistics are correct and informative for tuning workload performance. 

## Actual Behavior

Take a look at following CAS statistics. 
The workload is FIO 16k sequential read. The initial state of Cache device is 100% free (0% occupancy). 
The CAS statistics show 98.3% read hit ratio, which is totally wrong, which is easily justified by looking into "Block Statistics". 
According to "Block Statistics", more than 35% IO are read from core directly. 
Besides, the iostat log is also showing the majority IO occurs on core device. 
[iostat.log](https://github.com/Open-CAS/open-cas-linux/files/6505433/VROC-CAS-device_16k_Seq_0write_IODepth128_Workers1_1_iostat.txt)

Cache Id                  1
Cache Size                97333716 [4KiB Blocks] / 371.30 [GiB]
Cache Device              /dev/nvme0n1
Core Devices              1
Inactive Core Devices     0
Write Policy              wb
Eviction Policy           lru
Cleaning Policy           alru
Promotion Policy          always
Cache line size           16 [KiB]
Metadata Memory Footprint 1.4 [GiB]
Dirty for                 0 [s] / Cache clean
Metadata Mode             normal
Status                    Running

╔══════════════════╤══════════╤══════╤═════════════╗
║ Usage statistics │  Count   │  %   │   Units     ║
╠══════════════════╪══════════╪══════╪═════════════╣
║ Occupancy        │ 97071504 │ 99.7 │ 4KiB Blocks ║
║ Free             │   262212 │  0.3 │ 4KiB Blocks ║
║ Clean            │ 97071504 │ 99.7 │ 4KiB Blocks ║
║ Dirty            │        0 │  0.0 │ 4KiB Blocks ║
╚══════════════════╧══════════╧══════╧═════════════╝

╔══════════════════════╤══════════╤═══════╤══════════╗
║ Request statistics   │  Count   │   %   │ Units    ║
╠══════════════════════╪══════════╪═══════╪══════════╣
║ Read hits            │ 44805984 │  98.3 │ Requests ║
║ Read partial misses  │        0 │   0.0 │ Requests ║
║ Read full misses     │   758385 │   1.7 │ Requests ║
║ Read total           │ 45564369 │ 100.0 │ Requests ║
╟──────────────────────┼──────────┼───────┼──────────╢
║ Write hits           │        0 │   0.0 │ Requests ║
║ Write partial misses │        0 │   0.0 │ Requests ║
║ Write full misses    │        0 │   0.0 │ Requests ║
║ Write total          │        0 │   0.0 │ Requests ║
╟──────────────────────┼──────────┼───────┼──────────╢
║ Pass-Through reads   │        0 │   0.0 │ Requests ║
║ Pass-Through writes  │        0 │   0.0 │ Requests ║
║ Serviced requests    │ 45564369 │ 100.0 │ Requests ║
╟──────────────────────┼──────────┼───────┼──────────╢
║ Total requests       │ 45564369 │ 100.0 │ Requests ║
╚══════════════════════╧══════════╧═══════╧══════════╝

╔══════════════════════════════════╤═══════════╤═══════╤═════════════╗
║ Block statistics                 │   Count   │   %   │   Units     ║
╠══════════════════════════════════╪═══════════╪═══════╪═════════════╣
║ Reads from core(s)               │  97071445 │ 100.0 │ 4KiB Blocks ║
║ Writes to core(s)                │         0 │   0.0 │ 4KiB Blocks ║
║ Total to/from core(s)            │  97071445 │ 100.0 │ 4KiB Blocks ║
╟──────────────────────────────────┼───────────┼───────┼─────────────╢
║ Reads from cache                 │ 179223936 │  64.9 │ 4KiB Blocks ║
║ Writes to cache                  │  97071474 │  35.1 │ 4KiB Blocks ║
║ Total to/from cache              │ 276295410 │ 100.0 │ 4KiB Blocks ║
╟──────────────────────────────────┼───────────┼───────┼─────────────╢
║ Reads from exported object(s)    │ 276295381 │ 100.0 │ 4KiB Blocks ║
║ Writes to exported object(s)     │         0 │   0.0 │ 4KiB Blocks ║
║ Total to/from exported object(s) │ 276295381 │ 100.0 │ 4KiB Blocks ║
╚══════════════════════════════════╧═══════════╧═══════╧═════════════╝

## Steps to Reproduce

1. Start a new CAS instance (16k cache line size, wb mode) on a NVMe drive (eg. Intel P5800X)
2. Attach a core device (eg. Intel S4510)
3. Run FIO 16k sequential read test, restrict the fio "--size" to be less than the capacity of cache device
4. Check CAS statistics (# casadm -P -i 1) after the test

## Your Environment

* OpenCAS version (commit hash or tag):   v21.3
* Operating System:  RHEL7.8
* Kernel version:  3.10.0-1127.el7.x86_64
* Cache device type (NAND/Optane/other): Intel P5800X 400GB
* Core device type (HDD/SSD/other):  Intel S4510 1.92TB
* Cache configuration:  
* * Cache mode: (default: wt)  wb
* * Cache line size: (default: 4)  16
* * Promotion policy: (default: always) always
* * Cleaning policy: (default: alru)  alru
* * Sequential cutoff policy: (default: full) full
* Other (e.g. _lsblk_, _casadm -P_, _casadm -L_)
# casadm -V
╔═════════════════════════╤══════════════════════╗
║ Name                    │       Version        ║
╠═════════════════════════╪══════════════════════╣
║ CAS Cache Kernel Module │ 21.03.0.0496.release ║
║ CAS Disk Kernel Module  │ 21.03.0.0496.release ║
║ CAS CLI Utility         │ 21.03.0.0496.release ║
╚═════════════════════════╧══════════════════════╝

# casadm -L
type    id   disk           status    write policy   device
cache   1    /dev/nvme0n1   Running   wb             -
└core   1    /dev/sdc       Active    -              /dev/cas1-1

**Comment by Ostrokrzew** [comment]
Hi, I tried to investigate the problem but I can't reproduce it.
Can I ask you for the exact *fio* setup you used? Logs from *iostat* look like the *fio* size was greater than the cache volume size.

**Comment by eriksun421** [comment]
Here is the fio command. 

numactl --cpunodebind=0 fio --ioengine=libaio --direct=1 --buffered=0 --norandommap --randrepeat=0 --thread --unified_rw_reporting=1 --group_reporting --do_verify=0 --refill_buffers --stonewall --disable_slat=1 --disable_lat=1 --disable_bw=1 --cpus_allowed_policy=split --cpus_allowed=0-35 --time_based --runtime=600 --ramp_time=600 --name=VROC-CAS-device_16k_Seq_0write_IODepth128_Workers1_1 --size=379185M --filename=/dev/cas1-1 --rw=read --bs=16k --iodepth=128 --iodepth_batch=128 --iodepth_batch_complete_min=128 --numjobs=1 --minimal --eta=never --output=./2021_0518_1832_VROC-CAS-device/VROC-CAS-device_16k_Seq_0write_IODepth128_Workers1_1.csv

**Comment by Deixx** [comment]
Hi,

the iostat logs show IO on backend device for most of the time, but there's a moment in the logs after which IO switches to cache device. The most probable cause of this behavior seems using the `ramp_time` parameter, which results in running a warm-up workload before the actual fio measurements start. That workload is cached and included in CAS statistics. When the proper workload starts, it accesses data already cached by the ramp-up workload.

**Comment by Ostrokrzew** [comment]
Your *iostat* logs show reads from cache device since line **880** - there was some I/O traffic - and no reads from core device since **889** line (next measurement).
It doesn't look like a bug. More like a wrong *fio* configuration for this test. Manual says that time-based *fio* would loop over the same workload as many times as the `runtime` allows. And as Deixx said, `ramp_time` may also have an effect.

**Comment by eriksun421** [comment]
Can you please explain why more than 35% IO 4k blocks are read from core device when you look at the block statistics ?

**Comment by Ostrokrzew** [comment]
Your cache device is so fast that in the last 3.5 minute of this *fio* workload, cached data is read from cache device almost two more times.
Data from iostat.log:
reads from core device per second: ca. 780 requests or 390 MB (first 16.5 minutes)
reads from cache device per second: ca. 195 000 requests or 3 GB (last 3.5 minutes only)

195 000 / 780 = 250 times faster
3GB / 390 MB ≈ 8 times faster

So ca. 35% of the blocks read from the core device are read during the first 16.5 minutes of the *fio*.

**Comment by Ostrokrzew** [comment]
Can I close the issue?

**Comment by eriksun421** [comment]
This issue is about the read hit statistics, which I think has nothing to do with how fast I/O is read from cache/core devices. 
Just do a simple math calculation, there are 758385 read full miss requests. Each request is equal to 16kiB as fio block size is 16k.
758385 *4 = 3033540 4KiB blocks, which is not aligned with 97071445 (4KiB blocks read from core).

**Comment by Ostrokrzew** [comment]
You found a very interesting bug. Ignoring the extra entries from *Udev* (can be turned off for a test), the miss statistics are divided by the number of physical cores in the CPU (on my machine).
The command I used (slightly modified):
```bash
numactl --cpunodebind=0 fio --ioengine=libaio --direct=1 --buffered=0 --norandommap --randrepeat=0 --thread --unified_rw_reporting=1 --group_reporting --do_verify=0 --refill_buffers --stonewall --disable_slat=1 --disable_lat=1 --disable_bw=1 --cpus_allowed_policy=split --cpus_allowed=0-15 --name=VROC-CAS-device_16k_Seq_0write_IODepth128_Workers1_1 --size=2000M --filename=/dev/cas1-1 --rw=read --bs=16k --iodepth=128 --iodepth_batch=128 --iodepth_batch_complete_min=128 --numjobs=1 --minimal --eta=always --output=./VROC-CAS-device_16k_Seq_0write_IODepth128_Workers1_1.csv
```
***
Stats after *fio*:
```
[root@localhost ~]# casadm -P -i 1                           
Cache Id                  1                                  
Cache Size                5204384 [4KiB Blocks] / 19.85 [GiB]
Cache Device              /dev/nvme0n1p1                     
Core Devices              1                                  
Inactive Core Devices     0                                  
Write Policy              wb                                 
Eviction Policy           lru                                
Cleaning Policy           alru                               
Promotion Policy          always                             
Cache line size           8 [KiB]                            
Metadata Memory Footprint 159.6 [MiB]                        
Dirty for                 0 [s] / Cache clean                
Metadata Mode             normal                             
Status                    Running                            
                                                             
+==================+=========+======+=============+          
| Usage statistics |  Count  |  %   |   Units     |
+==================+=========+======+=============+
| Occupancy        |  512000 |  9.8 | 4KiB Blocks |
| Free             | 4692384 | 90.2 | 4KiB Blocks |
| Clean            |  512000 |  9.8 | 4KiB Blocks |
| Dirty            |       0 |  0.0 | 4KiB Blocks |
+==================+=========+======+=============+

+======================+=======+=======+==========+
| Request statistics   | Count |   %   | Units    |
+======================+=======+=======+==========+
| Read hits            |     0 |   0.0 | Requests |
| Read partial misses  |     0 |   0.0 | Requests |
| Read full misses     | 16000 | 100.0 | Requests |
| Read total           | 16000 | 100.0 | Requests |
+----------------------+-------+-------+----------+
| Write hits           |     0 |   0.0 | Requests |
| Write partial misses |     0 |   0.0 | Requests |
| Write full misses    |     0 |   0.0 | Requests |
| Write total          |     0 |   0.0 | Requests |
+----------------------+-------+-------+----------+
| Pass-Through reads   |     0 |   0.0 | Requests |
| Pass-Through writes  |     0 |   0.0 | Requests |
| Serviced requests    | 16000 | 100.0 | Requests |
+----------------------+-------+-------+----------+
| Total requests       | 16000 | 100.0 | Requests |
+======================+=======+=======+==========+

+==================================+========+=======+=============+
| Block statistics                 | Count  |   %   |   Units     |
+==================================+========+=======+=============+
| Reads from core(s)               | 512000 | 100.0 | 4KiB Blocks |
| Writes to core(s)                |      0 |   0.0 | 4KiB Blocks |
| Total to/from core(s)            | 512000 | 100.0 | 4KiB Blocks |
+----------------------------------+--------+-------+-------------+
| Reads from cache                 |      0 |   0.0 | 4KiB Blocks |
| Writes to cache                  | 512000 | 100.0 | 4KiB Blocks |
| Total to/from cache              | 512000 | 100.0 | 4KiB Blocks |
+----------------------------------+--------+-------+-------------+
| Reads from exported object(s)    | 512000 | 100.0 | 4KiB Blocks |
| Writes to exported object(s)     |      0 |   0.0 | 4KiB Blocks |
| Total to/from exported object(s) | 512000 | 100.0 | 4KiB Blocks |
+==================================+========+=======+=============+

+====================+=======+=====+==========+
| Error statistics   | Count |  %  | Units    |
+====================+=======+=====+==========+
| Cache read errors  |     0 | 0.0 | Requests |
| Cache write errors |     0 | 0.0 | Requests |
| Cache total errors |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Core read errors   |     0 | 0.0 | Requests |
| Core write errors  |     0 | 0.0 | Requests |
| Core total errors  |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Total errors       |     0 | 0.0 | Requests |
+====================+=======+=====+==========+
```
***
`lscpu -e` output
```
CPU NODE SOCKET CORE L1d:L1i:L2:L3 ONLINE MAXMHZ    MINMHZ
0   0    0      0    0:0:0:0       yes    2700.0000 800.0000
1   0    0      1    1:1:1:0       yes    2700.0000 800.0000
2   0    0      2    2:2:2:0       yes    2700.0000 800.0000
3   0    0      3    3:3:3:0       yes    2700.0000 800.0000
4   0    0      4    4:4:4:0       yes    2700.0000 800.0000
5   0    0      5    5:5:5:0       yes    2700.0000 800.0000
6   0    0      6    6:6:6:0       yes    2700.0000 800.0000
7   0    0      7    7:7:7:0       yes    2700.0000 800.0000
8   0    0      0    0:0:0:0       yes    2700.0000 800.0000
9   0    0      1    1:1:1:0       yes    2700.0000 800.0000
10  0    0      2    2:2:2:0       yes    2700.0000 800.0000
11  0    0      3    3:3:3:0       yes    2700.0000 800.0000
12  0    0      4    4:4:4:0       yes    2700.0000 800.0000
13  0    0      5    5:5:5:0       yes    2700.0000 800.0000
14  0    0      6    6:6:6:0       yes    2700.0000 800.0000
15  0    0      7    7:7:7:0       yes    2700.0000 800.0000
```
***
Calculations:

**unify the unit to 4KiB blocks:**
16 000 * 16 / 4 = 64 000

**multiply the result by the number of processor cores:**
64 000 * 8 = 512 000
***
Why is it like that? I do not know. I am going to ask someone from the team for help, as I cannot handle this issue anymore.

**Comment by mmichal10** [comment]
Since the issue was fixed by b82b338c53a5840611e607bbecca7755a7467b7f, it affects only v21.3 branch
